### PR TITLE
PXC-2687: BF abort of a SP causes a write-after-free error

### PR DESCRIPTION
**Pull request by voodam**
https://jira.percona.com/browse/PXC-2687

Extend PXC-3226: disable wsrep_retry_autocommit mechanism for all required commands inside CALL statement.

Bug is not reproduced in 8.0.

### CODE REVIEW

File: sql/sys_vars.cc
Comment (kamil-holubicki) [inline sql/sys_vars.cc:6410]: Why this change is needed?

**Review by kamil-holubicki — COMMENTED** [review]
Is it possible to create MTR test for this?

File: sql/sql_class.cc
Comment (kamil-holubicki) [inline sql/sql_class.cc]: Is this OK? `start_result_metadata` returns false in case of success.
So `is_result_set_started` will not be set in success path.

File: sql/sys_vars.cc
Comment (voodam) [reply sql/sys_vars.cc:6410]: For preventing memory leak.

File: sql/sql_class.cc
Comment (voodam) [reply sql/sql_class.cc]: No, it's wrong, fixed.

File: sql/sys_vars.cc
Comment (voodam) [reply sql/sys_vars.cc:6410]: If `PREALLOCATED` is set, variable will be freed, this is configured in that way for other `wsrep_*` string variables.

**Comment by venkatesh-prasad-v** [comment]
Hi! Is there any way to reproduce the `write-after-free` error? Can you please create an MTR test for this?

**Comment by voodam** [comment]
> Hi! Is there any way to reproduce the `write-after-free` error? Can you please create an MTR test for this?

Hi! Probably yes, I can see some testcase examples, where ASAN is mentioned, will try soon.

**Comment by voodam** [comment]
Finally added test.

File: scripts/wsrep_sst_common.sh
Comment (venkatesh-prasad-v) [inline scripts/wsrep_sst_common.sh]: This is still not guaranteed to work, as asan can report memory errors in bash itself.
```
2022-06-29T04:10:27.351633Z 0 [Note] WSREP: Initiating SST/IST transfer on JOINER side (wsrep_sst_xtrabackup-v2 --role 'joiner' --address '127.0.0.1:13009' --datadir '/home/venki/work/pxc/57/asan/mysql-test/var/mysqld.2/data/' --defaults-file '/home/venki/work/pxc/57/asan/mysql-test/var/my.cnf' --defaults-group-suffix '.2' --parent '988760' --mysqld-version '5.7.38-41'   '' )

=================================================================
==988775==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 2 byte(s) in 1 object(s) allocated from:
    #0 0x7f9b53747808 in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cc:144
    #1 0x555fa03974f3 in xmalloc (/usr/bin/bash+0x904f3)

SUMMARY: AddressSanitizer: 2 byte(s) leaked in 1 allocation(s).

=================================================================
==988779==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 2 byte(s) in 1 object(s) allocated from:
    #0 0x7f9b53747808 in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cc:144
    #1 0x555fa03974f3 in xmalloc (/usr/bin/bash+0x904f3)

SUMMARY: AddressSanitizer: 2 byte(s) leaked in 1 allocation(s).
```

In my opinion, it is better to move the unsetting logic to server. The below patch worked well locally.
```diff
diff --git a/sql/wsrep_sst.cc b/sql/wsrep_sst.cc
index 3089984e40c..c9731a89acb 100644
--- a/sql/wsrep_sst.cc
+++ b/sql/wsrep_sst.cc
@@ -780,6 +780,16 @@ static void* sst_joiner_thread (void* a)
 
 #define WSREP_SST_AUTH_ENV "WSREP_SST_OPT_AUTH"
 
+#if defined(HAVE_ASAN)
+/*
+ * When MySQL is built with ASAN, libasan will be propagated to child processes
+ * via LD_PRELOAD. This can be annoying because unix utils including bash can
+ * have memory errors itself. Let's unset LD_PRELOAD on ASAN builds.
+ */
+static void reset_ld_preload(wsp::env& env) {
+  env.append("LD_PRELOAD=");
+}
+#endif
 static int sst_append_auth_env(wsp::env& env, const char* sst_auth)
 {
   int const sst_auth_size= strlen(WSREP_SST_AUTH_ENV) + 1 /* = */
@@ -857,6 +867,10 @@ static ssize_t sst_prepare_other (const char*  method,
     return -env.error();
   }
 
+#if defined(HAVE_ASAN)
+  reset_ld_preload(env);
+#endif
+
   if ((ret= sst_append_auth_env(env, sst_auth)))
   {
     WSREP_ERROR("sst_prepare_other(): appending auth failed: %d", ret);
@@ -1595,6 +1609,10 @@ wsrep_cb_status_t wsrep_sst_donate_cb (void* app_ctx, void* recv_ctx,
     return WSREP_CB_FAILURE;
   }
 
+#if defined(HAVE_ASAN)
+  reset_ld_preload(env);
+#endif
+
   /* Wait for wsrep-SE to initialize that also signals
   completion of init_server_component which is important before we initiate
   any meangiful action especially DONOR action from this node. */
```

File: mysql-test/suite/galera/t/bug2687.test
Comment (venkatesh-prasad-v) [inline mysql-test/suite/galera/t/bug2687.test]: Considering the test case coverage, I would prefer running this test irrespective of asan builds. Can you remove this `include/asan` line and other related require files?

File: mysql-test/suite/galera/t/bug2687.test
Comment (venkatesh-prasad-v) [inline mysql-test/suite/galera/t/bug2687.test]: Is it not posisble to simulate BF abort with debug syncs?

**Review by venkatesh-prasad-v — CHANGES_REQUESTED** [review]

File: scripts/wsrep_sst_common.sh
Comment (kamil-holubicki) [reply scripts/wsrep_sst_common.sh]: This is a good idea.
We have a similar workaround in 8.0, but in Jenkins script
https://github.com/kamil-holubicki/jenkins-pipelines/blob/parallel-mtr/pxc/local/test-binary-pxc-parallel-mtr#L148
I would say, the one proposed by @venkatesh-prasad-v is better, it is compiled-in only in test builds, and works locally without the magic we currently do in 8.0 Jenkins. My proposal is to propagate it to 8.0 as well. Then I will remove the workaround from Jenkins script.

File: mysql-test/suite/galera/t/bug2687.test
Comment (voodam) [reply mysql-test/suite/galera/t/bug2687.test]: I don't know the point in code where to set debug sync, so it can take some time. I'll try to do this quickly.

**Comment by voodam** [comment]
@venkatesh-prasad-v, 

I mentioned that "malformed packet" error (`--error 2027`) is produced exactly because of not sending metadata. It produced in another valid case, here: https://github.com/percona/percona-xtradb-cluster/blob/5.7/mysql-test/suite/galera/t/galera_sp_bf_abort2.test#L22 if we add `SELECT f1 FROM t1`, "malformed packet" error will occur too if `wsrep_retry_autocommit = 1`.

So, why we don't need to send metadata? Git blame didn't give much: https://github.com/percona/percona-xtradb-cluster/commit/facd915a4673#diff-65c68b03ca220f14939f817d3c9cfe7767580750f9ef47876b383780a0fd82c3R2979. I tried to run `galera`, `galera_3nodes` and `wsrep` test suites (are these all cluster-specific tests?) and all of them wasn't affected by deleting `return FALSE` line (and hence sending metadata).

**Review by kamil-holubicki — COMMENTED** [review]
As this is 5.7 fix, It has to be done according to GCA merge process

**Comment by voodam** [comment]
> As this is 5.7 fix, It has to be done according to GCA merge process

Done

**Comment by voodam** [comment]
@kamil-holubicki, this PR includes patch from [PS-8327](https://github.com/percona/percona-server/pull/4839). What is better: full copy-paste PS-8327 to PXC-2687 or leave as is until merging percona-server to PXC?

**Comment by kamil-holubicki** [comment]
> @kamil-holubicki, this PR includes patch from [PS-8327](https://github.com/percona/percona-server/pull/4839). What is better: full copy-paste PS-8327 to PXC-2687 or leave as is until merging percona-server to PXC?

I see... As there were no complains about PXC-2687 I would wait for PS to be merged into PXC and then add what is necessary from PXC point of view. If we do the other way, there will be conflicts during the merge.

**Comment by kamil-holubicki** [comment]
Hi @voodam , we merged PS 5.7.40 into the PXC, so the above-mentioned PS fix is already there. Could you please check this PR? (it shows some conflicts)

**Comment by voodam** [comment]
> Hi @voodam , we merged PS 5.7.40 into the PXC, so the above-mentioned PS fix is already there. Could you please check this PR? (it shows some conflicts)

Hi! Now it should be OK.

UPD. Probably some problems after the merge: 
```
storage/innobase/handler/ha_innodb.cc:9508:12: error: unused variable 'err' [-Werror,-Wunused-variable]
                        dberr_t err = row_lock_table_for_mysql(
```

**Comment by voodam** [comment]
Catched a bug due the broken galera_sp_bf_abort2 test. We need to require that all instructions should retry in autocommit to retry CALL statement, so I fixed that.

**Comment by voodam** [comment]
https://pxc.cd.percona.com/view/PXC%205.7/job/pxc-5.7-pipeline/5149/#showFailuresLink
Now all is OK, `merge` and `merge_mmap` were broken somewhere in the previous commits.

@kamil-holubicki I can't to sign CLA for some reason: it redirects me to GitHub, but stays pending.

**Review by kamil-holubicki — APPROVED** [review]
Looks fine. Please squash commits and write descriptive commit message (what was the problem, why was the problem, how it was solved)

**Comment by voodam** [comment]
Done and also rewinded to GCA.
UPD. Some problems with branches, I'll fix it a little bit later.
UPD2. I have a trouble with following GCA process, waiting for a responce.

**Comment by voodam** [comment]
Rewinded back to 5.7, GCA process was cancelled.

**Comment by voodam** [comment]
https://pxc.cd.percona.com/view/PXC%205.7/job/pxc-5.7-pipeline/5167/

File: sql/sp_instr.h
Comment (venkatesh-prasad-v) [inline sql/sp_instr.h]: This must be inside `#ifdef WITH_WSREP`

**Review by venkatesh-prasad-v — CHANGES_REQUESTED** [review]

File: sql/sp_instr.h
Comment (kamil-holubicki) [reply sql/sp_instr.h]: Why? I would rather say: "please revert formating change" :)
This code is from PS

Comment (venkatesh-prasad-v) [reply sql/sp_instr.h]: Sorry for mistakenly commenting on `get_command()`. I intended to make the below `get_lex()` function be enclosed under `#ifdef WITH_WSREP` as this change is a PXC specific change.

Comment (voodam) [reply sql/sp_instr.h]: Done (get_lex I suppose, get_command is just reformatted).

Comment (venkatesh-prasad-v) [reply sql/sp_instr.h]: Can you please revert the formatting changes done for the `get_command()` function? This will cause merge conflict in future when the nearby lines are changed in PS.

```suggestion
  int get_command() const
  { 
    return m_lex ? m_lex->sql_command : -1; 
  }
```

Comment (voodam) [reply sql/sp_instr.h]: Done

**Review by venkatesh-prasad-v — APPROVED** [review]

**Pull request merged**
